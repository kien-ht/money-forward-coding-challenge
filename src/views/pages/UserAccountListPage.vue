<template>
  <MainLayout>
    <h1 class="user-name">{{ currentUserName }}</h1>
    <b-table striped small :fields="fields" :items="items" responsive="sm">
      <template #cell(id)="{ item }">
        {{ item.id }}
      </template>

      <template #cell(name)="{ item }">
        <b class="text-info">{{ item.name }}</b>
      </template>

      <template #cell(balance)="{ item }">
        {{ item.balance | formatCurrency }}
      </template>
    </b-table>
    <div class="footer-area">
      <b-button size="md" variant="dark" @click="onClickBackButton"
        >Back</b-button
      >
    </div>
  </MainLayout>
</template>

<script land="ts">
import Vue from "vue";
import MainLayout from "@/views/layouts/MainLayout.vue";

export default Vue.extend({
  name: "UserAccountListPage",

  components: {
    MainLayout,
  },

  filters: {
    formatCurrency(money) {
      return money.toLocaleString("de-DE");
    },
  },

  data() {
    return {
      fields: [
        { key: "id", label: "Account Id" },
        { key: "name", label: "Name" },
        { key: "balance", label: "Balance" },
      ],
      items: [],
      currentUserName: "",
    };
  },

  computed: {
    currentUserId() {
      return this.$route.params.id;
    },
  },

  created() {
    this.fetchData();
  },

  methods: {
    // region [EVENT]
    onClickBackButton() {
      this.$router.go(-1);
    },
    // end region

    // region [TRANSACTION]
    async fetchUser() {
      const response = await this.$api.userService.getUser(this.currentUserId);

      this.currentUserName = response ? response.attributes.name : "";
    },

    async fetchData() {
      this.fetchUser();

      if (!this.$store.getters.currentSearchUserAccountList.length) {
        await this.$store.dispatch(
          "setCurrentSearchUserAccountList",
          this.currentUserId
        );
      }

      this.items = this.$store.getters.currentSearchUserAccountList;
    },
    // end region
  },
});
</script>

<style lang="scss" scoped>
.user-name {
  text-align: left;
  text-transform: uppercase;
}
.footer-area {
  position: absolute;
  bottom: 0;
}
</style>
