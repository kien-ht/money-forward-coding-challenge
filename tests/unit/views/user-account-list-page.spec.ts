import { shallowMount, mount } from "@vue/test-utils";
import UserAccountListPage from "@/views/pages/UserAccountListPage.vue";

describe("UserAccountListPage.vue", () => {
  it("snapshot", () => {
    const wrapper = mount(UserAccountListPage, {
      stubs: ["router-link", "router-view"],
    });
    expect(wrapper).toMatchSnapshot();
  });

  let wrapper: any;

  beforeEach(() => {
    wrapper = shallowMount(UserAccountListPage, {
      data: () => ({ currentUserName: "test user name" }),
    });
  });

  it("renders user name", () => {
    const userName = wrapper.find("h1").text();
    expect(userName).toBe("test user name");
  });
});
