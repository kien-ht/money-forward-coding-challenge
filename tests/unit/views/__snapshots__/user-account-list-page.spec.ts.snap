// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`UserAccountListPage.vue snapshot 1`] = `
<div class="layout-wrapper">
  <div class="page-topbar bg-white">
    <router-link-stub to="/"><span><img src="@/assets/money-forward-logo.png" alt="" height="50"></span></router-link-stub>
  </div>
  <div class="h-100 container-fluid">
    <div class="row h-100">
      <div class="bg-warning py-5 px-4 d-none d-md-block col-3">
        <ul class="nav nav-pills flex-column"><button type="button" class="btn w-100 btn-dark">Account Search</button></ul>
      </div>
      <div class="p-5 col">
        <h1 class="user-name"></h1>
        <div class="table-responsive-sm">
          <table role="table" aria-busy="false" aria-colcount="3" class="table b-table table-striped table-sm">
            <!---->
            <!---->
            <thead role="rowgroup" class="">
              <!---->
              <tr role="row" class="">
                <th role="columnheader" scope="col" aria-colindex="1" class="">
                  <div>Account Id</div>
                </th>
                <th role="columnheader" scope="col" aria-colindex="2" class="">
                  <div>Name</div>
                </th>
                <th role="columnheader" scope="col" aria-colindex="3" class="">
                  <div>Balance</div>
                </th>
              </tr>
            </thead>
            <tbody role="rowgroup"></tbody>
            <!---->
          </table>
        </div>
        <div class="footer-area"><button type="button" class="btn btn-dark btn-md">Back</button></div>
      </div>
    </div>
  </div>
</div>
`;
